### ISSUE TITLE
CMAQ exposures API - Plans for expanding data

### ISSUE THREAD

**Issue by lstillwe** [ResearchSoftwareInstitute/greendatatranslator#132]
As per discussion on 4/11/18, plans are as follows:

Temporal expansion of existing Exposures Service/API - Explore whether we can obtain additional years of data on CMAQ exposures (beyond 2010, 2011) from the US EPA [Sarav]
Spatial expansion of existing Exposures Service/API - Generate higher-resolution CMAQ estimates - one year likely by end of feasibility assessment [Sarav's student]
Pollutant expansion of existing Exposures Service/API - Incorporate additional exposure estimates for 2010 and 2011 (not just PM2.5, ozone) [Lisa]
Data quality metrics for Exposures Service/API - Stick with current metrics for PM2.5 and ozone (2010, 2011), at least for the time being [Lisa]
R-LINE - hold off on this effort, at least for now [Sarav]

@karafecho Are there any other comment you wish to preserve from #123 ?

**Comment by karafecho** [comment]
As per discussion on 4/30/2018 and subsequent summary provided by Sarav, revised plans are:

1. **National-scale CMAQ data for longer time-series:** Sarav was able to do some research, and found an EPA project that has fused multiple years of CMAQ data with AQS-based observations from 2002 - 2014, and with additional years from 2015 - 2017 coming later in the next couple of years. Sarav was able to confirm acquisition of these for our use from the main EPA POC. IE will process and support multiple years of CMAQ data from 2002 - 2015, with focus on daily ave PM25 and MDA8 O3 from  12-km resolution modeling for the entire U.S. While 2002 - 2014 data are available right away from the EPA, 2015 data will be available later this year. The beauty of this new dataset is that they are all self-consistent and we are able to offer this longer time-series of exposures data, and make it appealing to the broader DT team beyond UNC. The downside is that we only have these 2 pollutants for the longer time-series.  Sarav will continue to explore if we can get hold of the raw CMAQ outputs from the EPA which may have the 100+ pollutants. For e.g. Maureen H may like to get hold of Formaldehyde predictions for her Fanconi Anemia end point, and there may be other use cases like this.

CORRECTION: Data are at the resolution of the US Census tract

NOTE: Original CMAQ data are at 36-km resolution for 2010 and 12-km resolution for 2011

NOTE: File sizes of raw NetCDF files for original CMAQ data are 40Gbs for 2010 and 544 Gbs for 2011; total row count for the 2010, 2011 CMAQ DB is 1,347,436,920 (that's right - 1.3 billion!!)

2. **High resolution R-LINE for Traffic-related Emissions:**  IE will explore creating high resolution census block level national-scale primary PM2.5 due to traffic-related sources for a single year. We will compare / contrast our source-based modeling with RENCI/Lisa's work that uses DOT/FHWA road activity-based regression approach.

**Comment by karafecho** [comment]
Also see #20 .

**Comment by karafecho** [comment]
Per Sarav, 6/13/2018, re new CMAQ data noted above: "Since this is higher spatial resolution, and uniform for a longer time period, this will all be in a new format.  Assume CSV files that are geocoded by Census tract centroids, and one value per day." Plans are to send new CMAQ estimates for year 2011, with comparisons to previous CMAQ estimates, by end of June. Sample data will be send to Lisa in the meantime.

**Comment by lstillwe** [comment]
Actually I have not received any of the new CMAQ data from Sarav yet. I have received Roadway data - but no new CMAQ data.

**Comment by karafecho** [comment]
From Sarav, 7/13/2018:

2010 had no fires. But 2011 was a severe fire year, espl in Eastern NC, and on some days the smoke plume traveled to the Triangle region.

**Comment by arunacs** [comment]
See #141

**Comment by karafecho** [comment]
**Per discussion with Sarav, Adel, Dongmei, Alex, Steve, Lisa, Kara on 7/25/2018:**

**Deliverables:**

- I.E. will deliver 2011-2015 CMAQ data by mid September. The specific deliverable will be nationwide (continental US) CMAQ estimates of PM2.5 and ozone, provided at a temporal resolution of 24-hour periods (daily, not hourly estimates) and a spatial resolution of the US Census tract. 2011 CMAQ estimates have been validated and were provided to Green Team on 7/24/18. IE is double checking the viability of delivering 2015 CMAQ estimates by mid September.
- IE will deliver 2002-2015 CMAQ data (daily, nationwide, US Census tract-level estimates of PM2.5 and ozone) by the end of the feasibility period. 
- We'd like to discuss the feasibility of a phased approach to give us time to incorporate the data prior to the September hackathon.

**Dongmei will:**

- Lead the effort to deliver this data.
- ​Participate in Translator scrums to collaborate with the team. (I've sent her an invitation.)
- Get us her GitHub handle to allow her to participate in the team.​

**Comment by arunacs** [comment]
Checked on 2015 data. EPA is still working on it, and expected to be available by the end of this year.

**Comment by lstillwe** [comment]
Setup new VM (bdt-cmaq.renci.org) and created PostgreSQL DB for this new Data/Service.
Wrote script to merge data files for pm2.5 and o3 and then successfully loaded data into DB - took less that 10 minutes for year 2011.

**Comment by lstillwe** [comment]
@karafecho FYI:
@arunacs I have noticed that the new CMAQ data for o3 and pm25 contains the daily average for pm25 and the daily maximum for for o3. Is it possible to get the max and average for each variable type with this new data? Otherwise I can not really support aggregation types for both variables.

**Comment by karafecho** [comment]
@lstillwe : @xu-hao : James is ready to integrate the 2012 and 2013 clinical data with the CMAQ, ACS, and roadway data. Perhaps one of you can add the 2012 and 2013 CMAQ data to Rockfish and let James know where he can find it? Thanks!

**Comment by lstillwe** [comment]
The new expanded CMAQ API is completed and now available here:http://bdt-cmaq.renci.org:8080/cmaq_exposures_api/v1/swagger.json
Also please see this repository for documentation: https://github.com/lstillwe/datatranslator-exposure-apis

**Comment by karafecho** [comment]
@lstillwe : You rock! Thanks.

**Comment by xu-hao** [comment]
eta for integrating fihr data oct/14

**Comment by karafecho** [comment]
@cpschmitt : See info above. May be helpful to you.

**Comment by karafecho** [comment]
@lstillwe : When you get a chance, perhaps you can look into the US EPA conUS CMAQ data and see if we can pull data on additional chemicals? It would be great to expand the 2002-2014 conUS CMAQ data to include data on a variety of chemicals, not just PM2.5 and ozone.

**Comment by lstillwe** [comment]
@karafecho I do have the new data from @arunacs however, I wanted to wait for an update from @arunacs to see if we can get error values for these new pollutants. (see https://github.com/ResearchSoftwareInstitute/greendatatranslator/issues/165#issuecomment-477848061). It requires a bit of work to add new columns to the existing table, so I would rather have all data in hand before taking on this task. If you think that the standard errors are not important at this time, I will go ahead and add the new exposure values as is.

**Comment by karafecho** [comment]
@lstillwe : Yes, please coordinate with @aruncas. If the standard errors will require a lot of work to obtain and add to the table, then perhaps we should move forward as is. WRT the new pollutants, my understanding was that these are for year 2010 only. If so, I'm wondering if we can obtain additional years of data. We have users at NIEHS who are eager for data on additional pollutants.

**Comment by arunacs** [comment]
Yes, additional years of data are easier to get to than getting standard errors. I should be getting those to you this week, after another round of QA. The problem I saw earlier has been resolved.

**Comment by karafecho** [comment]
@arunacs : Do we have access to data on dichlorobenzene? Does conUS provide a list of available chemical exposures? Just wondering.

**Comment by lstillwe** [comment]
Just want to touch base with @arunacs and @karafecho: Do we now have standard errors for the additional exposure data provided? If not should I move ahead and just add these new pollutants without standard errors? Thanks!

**Comment by karafecho** [comment]
If I'm not mistaken, I think @arunacs is hoping to pull additional years of data. Let's hold off until he replies.

**Comment by karafecho** [comment]
From Sarav, Jan 29:

We have created daily averages at census block resolution for the following additional pollutants:

CO, NO, NO2, NOx, SO2, Formaldehyde, Acetaldehyde for the same time period.

Benzene, a key mobile source air toxic, however is not available continuously. It is on and off for some reason, and I need to investigate more.

From Sarav, Feb 6:

The resolution is census tracts - identical to what we gave last time for O3 and PM2.5 for ICEES, and the source is EPA. ( Note the correction to the email thread below, and not census block centroids )

EPA will be generating additional years of data, and the only year we can expect soon is 2015. 

And btw, I confirmed that the Census Tract definition is from the 2010 definition of the US Census, for all years of CMAQ data.

**Comment by karafecho** [comment]
@arunacs : Pehaps you can provide an update on the above comments? I think we'd like to move forward with or without benzene. We're about to generate new ICEES tables for 2010-2016, using the now validated CAMP FHIR/FHIR PIT pipeline, and it would be great to include the additional airborne toxicants.

**Comment by arunacs** [comment]
Confirming the following:

Have 2015 data for O3 and PM25 from the EPA. Will process and add them. 2016 is next on the EPA list, but  unlikely before next Hackathon.
Benzene is available only intermittently - for the years 2002, 2006, 2013 and 2014. But Formaldehyde and Acetaldehyde, along with CO, NO, NO2, NOx and SO2  are continuous. Will provide links to these extractions.

**Comment by karafecho** [comment]
Great news re the 2015 EPA data. Do you have an ETA for when you will be able to provision the new data for O3, PM25, acetaldehyde, CO, NO, NO2, NOx, and SO2 for years 2002-2015, plus benzene for the years that are available?

**Comment by arunacs** [comment]
I need to do some additional processing and QA before I hand them off. With a new hire (in progress), I hoping to turn these over in a month or sooner from now.

**Comment by karafecho** [comment]
@arunacs : Thanks for meeting earlier today. Please provide an update in GitHub, noting what conUS data have already been sent to Lisa, an ETA for the 2015 data, and an expected release data for the 2016 data. Thanks!

**Comment by karafecho** [comment]
@arunacs : I checked with Hao. He needs the 2015 conUS data by next Friday, 8/30, at the latest. Thanks!

**Comment by lstillwe** [comment]
Unless I have missed something, what I have so far is 2002-2014 data for ozone and pm2.5. 
Nothing new beyond that.

**Comment by karafecho** [comment]
@arunacs : Given the fact that the hackathon is less than three weeks ago, we really need to push the new data so that Lisa can updated the API and Hao can integrate the data with the new patient data files. Let's aim for an ETA of Friday, August 30, 2019 to complete the following tasks: (1) send Lisa the 2002-2014 data for the six new airborne pollutants; (2) send Lisa the 2015 data for all eight airborne pollutants.

**Comment by arunacs** [comment]
@karafecho @lstillwe We now have 2015 and 2016 data for O3 and PM2.5 - same format as before.

[sarav@dogwood-login2 FAQSD]$ pwd
/proj/ie/proj/NIH-DataTranslator/for_RENCI/CMAQ/FAQSD
[sarav@dogwood-login2 FAQSD]$ ls -c1 201[56]
2016:
2016_ozone_daily_8hour_maximum.txt
2016_pm25_daily_average.txt

2015:
2015_ozone_daily_8hour_maximum.txt
2015_pm25_daily_average.txt

2016 data is hot off of the press, and I would like to do further review of the 2016 data, but sending along in the interest of possibly using this for the next hackathon, while I do further analyses.

OTHER POLLUTANTS:
These have been compiled from what is available, and the header sample is shown below:
ID,FIPS,CLon,CLat,COL,ROW,GLon,GLat,Date,CO_ppbv,NO_ppbv,NO2_ppbv,NOX_ppbv,SO2_ppbv,ALD2_ppbv,FORM_ppbv,BENZ_ppbv
36557,1001020100,-86.49001,32.47718,296,80,-86.4288,32.49273,2014/01/01,150.04996,1.61578,7.28188,8.89766,0.77594,0.5998,0.41742,0.14948
36557,1001020100,-86.49001,32.47718,296,80,-86.4288,32.49273,2014/01/02,191.94937,2.0213,8.55927,10.58056,0.61'

ps: Note that only 2002, 2006, 2013, 2014 have BENZ data as mentioned earlier, while other pollutants are continuous from 2002 to 2014 as mentioned earlier.

[sarav@dogwood-login2 OtherPollutants]$ pwd
/proj/ie/proj/NIH-DataTranslator/for_RENCI/CMAQ/OtherPollutants
[sarav@dogwood-login2 OtherPollutants]$ ls -c1
cmaq_downscaling_12US1_459x299_2014.csv
cmaq_downscaling_12US1_459x299_2013.csv
cmaq_downscaling_12US1_459x299_2012.csv
cmaq_downscaling_12US1_459x299_2011.csv
cmaq_downscaling_12US1_459x299_2010.csv
cmaq_downscaling_12US1_459x299_2009.csv
cmaq_downscaling_12US1_459x299_2008.csv
cmaq_downscaling_12US1_459x299_2007.csv
cmaq_downscaling_12US1_459x299_2006.csv
cmaq_downscaling_12US1_459x299_2005.csv
cmaq_downscaling_12US1_459x299_2004.csv
cmaq_downscaling_12US1_459x299_2003.csv
cmaq_downscaling_12US1_459x299_2002.csv

**Comment by karafecho** [comment]
This is great! Thanks, Sarav.

@lstillwe : Do you think we can modify the API to allow access to the data prior to (or during) the hackathon?

@xu-hao : Perhaps you can coordinate with Lisa to add the additional airborne pollutants to the new ICEES tables? If you have questions, feel free to holler.

**Comment by lstillwe** [comment]
@arunacs 
Questions ... 
1) The pm2.5 & ozone datasets have different headers for the 2015,2016 data, are they the same meaning?
Example:
OLD:Date,FIPS,Longitude,Latitude,ozone_daily_8hour_maximum(ppb),ozone_daily_8hour_maximum_stderr(ppb)
NEW:Date,Loc_Label1,Latitude,Longitude,Prediction,SEpred

2) The pm2.5, ozone dataset has a 11 digit FIPS code and the other variables have a 10 & 11 digit FIPS codes. It appears that the FIPS code may have been treated as a number and not a string in the Other data dataset. Should I just prepend the 10 digit FIPS code in these datasets with a zero?
Example:
OZONE DATA FIPS: 01001020100
OTHER DATA FIPS: 1001020100

Thanks,
Lisa

**Comment by arunacs** [comment]
@lstillwe Confirming answers to both questions. Yes and Yes. Thanks Lisa.

**Comment by lstillwe** [comment]
@arunacs  - Question about the additional CMAQ pollutants:
The header for this file includes columns for CLon,CLat,GLon,GLat and also COL,ROW. What is the difference between these values for lat,lon and can I use them for the the real lat,lon values, or do I have to calculate them from COL,ROW?

This is of urgent interest to @karafecho and @xu-hao 

Thanks very much,
Lisa

@arunacs - I have assumed that CLon and CLat - refer to the Centroid of the Census Block and gone ahead with the processing of data for Hao

**Comment by karafecho** [comment]
@arunacs : is this something that Chuck Change can help with as part of the NCE, which began on 1/1/20?

**Comment by karafecho** [comment]
Related to issue #165.

**Comment by karafecho** [comment]
@lstillwe : I believe you transferred conUS data from 2002-2016 on PM2.5 and ozone to @xu-hao. Have you also transferred conUS data from 2002-2014 on the eight new airborne pollutants? If not, please do so as soon as possible.

@arunacs : Have you downloaded conUS data from 2015 and 2016 on the eight new airborne pollutants? If that's available, is that something that Charles Chang can help with? Lisa also needs someone to write script to merge the data on the new pollutants with data on PM2.5 and ozone. FYI: We have budgeted for Charles' time in Green Team's NCE from January 1 through March 31.

**Comment by lstillwe** [comment]
@karafecho I have given both items to @xu-hao

**Comment by karafecho** [comment]
@xu-hao : Just to confirm, you now have 2002-2016 conUS data for all ten airborne pollutants, correct?

**Comment by karafecho** [comment]
From Hao: i have 2002 - 2016 on ozone and pm2.5, 2002 - 2014 on the additional 8 pollutants (except that only 2002, 2006, 2013, 2014 have BENZ data)

**Comment by arunacs** [comment]
> @arunacs - Question about the additional CMAQ pollutants:
> The header for this file includes columns for CLon,CLat,GLon,GLat and also COL,ROW. What is the difference between these values for lat,lon and can I use them for the the real lat,lon values, or do I have to calculate them from COL,ROW?
> 
> This is of urgent interest to @karafecho and @xu-hao
> 
> Thanks very much,
> Lisa
> 
> @arunacs - I have assumed that CLon and CLat - refer to the Centroid of the Census Block and gone ahead with the processing of data for Hao

@lstillwe,  CLon and CLat refer to the Centroid of the Census tracts, while GLon and GLat refer to the SW corner of the 12x12km grid-cell they were extracted from.

**Comment by arunacs** [comment]
> @lstillwe : I believe you transferred conUS data from 2002-2016 on PM2.5 and ozone to @xu-hao. Have you also transferred conUS data from 2002-2014 on the eight new airborne pollutants? If not, please do so as soon as possible.
> 
> @arunacs : Have you downloaded conUS data from 2015 and 2016 on the eight new airborne pollutants? If that's available, is that something that Charles Chang can help with? Lisa also needs someone to write script to merge the data on the new pollutants with data on PM2.5 and ozone. FYI: We have budgeted for Charles' time in Green Team's NCE from January 1 through March 31.

@karafecho, I need to verify which of these additional pollutants are available - either with the EPA from their simulations, or from other work that IE has done. I will get back to you.

**Comment by karafecho** [comment]
Thanks, @arunacs!
